lib/puppet/pops/types/p_object_type_extension.rb

```diff
@@ -0,0 +1,216 @@
+    end
+    @parameters = by_name
+  end
+
+  def check_param(type_param, v)
+    if v == :default
+      raise Puppet::ParseError, "No value provided for required #{type_param.label}" unless type_param.value?
+      v = type_param.value
+    end
+    TypeAsserter.assert_instance_of(nil, type_param.type, v) { type_param.label }
+  end
+
+  # Return the parameter values as positional arguments with values that represent `default` as :default. The
+  # array is stripped from trailing :default values
+  # @return [Array] the parameter values
+  # @api private
+  def init_parameters
+    result = @base_type.type_parameters(true).values.map do |tp|
+      v = @parameters[tp.name]
+      tp.value? && tp.value == v ? :default : v
+    end
+    # Remove trailing defaults but avoid empty result. At least one parameter must
+    # be present, even it is :default
+    result.pop while result.size > 1 && result.last == :default
+    result
+  end
+
+  # @api private
+  def eql?(o)
+    super(o) && @base_type.eql?(o.base_type) && @parameters.eql?(o.parameters)
+  end
+
+  # @api private
+  def generalize
+    @base_type
+  end
+
+  # @api private
+  def hash
+    @base_type.hash ^ @parameters.hash
+  end
+
+  # @api private
+  def loader
+    @base_type.loader
+  end
+
+  # @api private
+  def check_self_recursion(originator)
+    @base_type.check_self_recursion(originator)
+  end
+
+  # @api private
+  def create(*args)
+    @base_type.create(*args)
+  end
+
+  # @api private
+  def instance?(o, guard = nil)
+    @base_type.instance?(o, guard) && test_instance?(o, guard)
+  end
+
+  # @api private
+  def new_function
+    @base_type.new_function
+  end
+
+  # @api private
+  def simple_name
+    @base_type.simple_name
+  end
+
+  protected
+
+  # Creates an array of type parameters from the attributes that matches the
+  # type parameters by name. Type parameters for which there is no matching attribute
+  # will have `nil` in their corresponding position on the array.
+  #
+  # @return [Array] array of values from instance that maps to type parameters
+  def parameters_from_instance(instance, type_parameters)
+    attrs = @base_type.attributes(true)
+    type_parameters.keys.map do |pn|
+      attr = attrs[pn]
+      attr.nil? ? nil : instance.send(pn)
+    end
+  end
+
+  # Checks that the given `param_values` hash contains all keys present in the `parameters` of
+  # this instance and that each keyed value is either assignable to the type stored in the
+  # `parameters` hash or equal to the value in that hash.
+  #
+  # This method is only called when a given type is found to be assignable to the base type of
+  # this extension.
+  #
+  # @param param_values[Hash] the parameter values of the assignable type
+  # @param guard[RecursionGuard] guard against endless recursion
+  # @return [Boolean] true or false to indicate assignability
+  # @api public
+  def test_assignable?(param_values, guard)
```

Comment:
Implementing case semantics here is probably a very good idea. Question is if that should be done both here and for the `test_instance?` method. This method compares a set of type_parameter values against another set of type_parameter values whereas the `test_instance?` will compare type parameter values against attribute values.

Reply:
If LHS is a Type then case matches with == instead of =~ IIRC (unsure if =~ is used when LHS is a Type and proposition is Type[T]). Think this is useful for comparing instance values vs. type parameters, not sure if it is correct for parameter / parameter matching.

Reply:
I see a commit that changes this. So... done.